document-store/src/main/java/org/hypertrace/core/documentstore/mongo/MongoCollection.java

```diff
@@ -386,28 +387,58 @@ public long total(Query query) {
   @Override
   public boolean bulkUpsert(Map<Key, Document> documents) {
     try {
-      List<UpdateOneModel<BasicDBObject>> bulkCollection = new ArrayList<>();
-      for (Entry<Key, Document> entry : documents.entrySet()) {
-        Key key = entry.getKey();
-        // insert or overwrite
-        bulkCollection.add(new UpdateOneModel<>(
-            this.selectionCriteriaForKey(key),
-            prepareUpsert(key, entry.getValue()),
-            new UpdateOptions().upsert(true)));
-      }
-
-      BulkWriteResult result = Failsafe.with(bulkWriteRetryPolicy)
-          .get(() -> collection.bulkWrite(bulkCollection, new BulkWriteOptions().ordered(false)));
+      BulkWriteResult result = bulkUpsertImpl(documents);
       LOGGER.debug(result.toString());
-
       return true;
-
     } catch (IOException | MongoServerException e) {
       LOGGER.error("Error during bulk upsert for documents:{}", documents, e);
       return false;
     }
   }
 
+  private BulkWriteResult bulkUpsertImpl(Map<Key, Document> documents) throws JsonProcessingException {
+    List<UpdateOneModel<BasicDBObject>> bulkCollection = new ArrayList<>();
+    for (Entry<Key, Document> entry : documents.entrySet()) {
+      Key key = entry.getKey();
+      // insert or overwrite
+      bulkCollection.add(new UpdateOneModel<>(
+          this.selectionCriteriaForKey(key),
+          prepareUpsert(key, entry.getValue()),
+          new UpdateOptions().upsert(true)));
+    }
+
+    return Failsafe.with(bulkWriteRetryPolicy)
+        .get(() -> collection.bulkWrite(bulkCollection, new BulkWriteOptions().ordered(false)));
+  }
+
+  @Override
+  public Iterator<Document> returnAndBulkUpsert(Map<Key, Document> documents) throws IOException {
```

Comment:
nit: rename to bulkUpsertAndReturnOlderDocuments?
returnAndBulkUpsert sounds a bit wierd.

Reply:
Done.